xds/internal/server/rds_handler_test.go

```diff
@@ -20,325 +20,391 @@ package server
-	if err != nil {
-		t.Fatalf("xdsClient.CancelRDS failed with error %v", err)
+	// Configure the management server with two route config resources.
+	routeResource1 := routeConfigResourceForName(route1)
+	routeResource2 := routeConfigResourceForName(route2)
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Routes:         []*v3routepb.RouteConfiguration{routeResource1, routeResource2},
+		SkipValidation: true,
 	}
-	if routeNameDeleted != route2 {
-		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v", routeNameDeleted, route2)
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
 	}
 
-	rdsUpdate := xdsresource.RouteConfigUpdate{}
-	// Invoke callback with the xds client with a certain route update. Due to
-	// this route update updating every route name that rds handler handles,
-	// this should write to the update channel to send to the listener.
-	fakeClient.InvokeWatchRouteConfigCallback(route1, rdsUpdate, nil)
-	rhuWant := map[string]xdsresource.RouteConfigUpdate{route1: rdsUpdate}
-	select {
-	case rhu := <-ch:
-		if diff := cmp.Diff(rhu.updates, rhuWant); diff != "" {
-			t.Fatalf("got unexpected route update, diff (-got, +want): %v", diff)
-		}
-	case <-ctx.Done():
-		t.Fatal("Timed out waiting for update from update channel.")
+	// Verify that the update is pushed to the handler's update channel.
+	wantUpdate := rdsHandlerUpdate{
+		updates: map[string]xdsresource.RouteConfigUpdate{
+			route1: defaultRouteConfigUpdate,
+			route2: defaultRouteConfigUpdate,
+		},
 	}
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
 
+	// Update the handler to watch only one of the two previous routes.
+	rh.updateRouteNamesToWatch(map[string]bool{route1: true})
+
+	// Verify that the other route is no longer requested.
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{route1})
+
+	// Verify that an update is pushed with only one route.
+	wantUpdate = rdsHandlerUpdate{updates: map[string]xdsresource.RouteConfigUpdate{route1: defaultRouteConfigUpdate}}
+	verifyUpdateFromChannel(ctx, t, updateCh, wantUpdate)
+
+	// Close the rds handler and verify that the watch is canceled.
 	rh.close()
-	routeNameDeleted, err = fakeClient.WaitForCancelRouteConfigWatch(ctx)
-	if err != nil {
-		t.Fatalf("xdsClient.CancelRDS failed with error: %v", err)
-	}
-	if routeNameDeleted != route1 {
-		t.Fatalf("xdsClient.CancelRDS called for route %v, want %v", routeNameDeleted, route1)
-	}
+	waitForResourceNames(ctx, t, resourceNamesCh, []string{})
 }
 
-// TestSuccessCaseTwoUpdatesAddAndDeleteRoute tests the case where the rds
+// Tests the case where the rds
 // handler receives an update with two routes, and then receives an update with
 // two routes, one previously there and one added (i.e. 12 -> 23). This should
 // cause the route that is no longer there to be deleted and cancelled, and the
 // route that was added should have a watch started for it.
```

Comment:
Wrap this comment to 80 characters please. Also optionally can you talk about the verifications after 23 is updated such as it doesn't send an update until 3 is received on management server etc.

Reply:
Done.